### PR TITLE
feat(redis_client): add logging support

### PR DESCRIPTION
**Pull request by felangel**
## Description

- feat(redis_client): add logging support
- feat: improve reconnection and command retry logic and simplify API

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/redis_client/example/main.dart [hunk 1]
```diff
@@ -9,24 +7,9 @@ Future<void> main() async {
...
-  // Get the value of a key.
-  final value = await client.get(key: key); // WORLD
-  assert(value == 'WORLD', 'Value should be "WORLD".');
-
-  // Delete the key.
-  await client.delete(key: key);
-
-  final finalValue = await client.get(key: key); // null
-  assert(finalValue == null, 'Key should not exist.');
+  // Send a command to the Redis server.
```
Comment (felangel) [inline packages/redis_client/example/main.dart]: Simplified the example

File: packages/redis_client/lib/src/redis_client.dart [hunk 2]
```diff
@@ -73,153 +141,169 @@ class RedisClient {
   /// A completer which completes when the client establishes a connection.
   var _connected = Completer<void>();
 
+  /// Whether the client is connected.
+  var _isConnected = false;
+
   /// A completer which completes when the client disconnects.
   /// Begins in a completed state since the client is initially disconnected.
   var _disconnected = Completer<void>()..complete();
```
Comment (eseidel) [inline packages/redis_client/lib/src/redis_client.dart:149]: I thought member variables were typically declared with types in Dart?

**Review by eseidel — COMMENTED** [review]
I'm too tired to give you a useful review, but I'm very glad to see this coming together!

File: packages/redis_client/lib/src/redis_client.dart (hunk 2 above)
Comment (felangel) [reply packages/redis_client/lib/src/redis_client.dart:149]: The lint rules we have consistently enforce omitting types if they can be inferred.

File: packages/redis_client/lib/src/redis_client.dart [hunk 3]
```diff
@@ -73,153 +141,169 @@ class RedisClient {
...
 
   /// A future which completes when the client establishes a connection.
-  Future<void> get connected => _connected.future;
+  Future<void> get _untilConnected => _connected.future;
```
Comment (bryanoltman) [inline packages/redis_client/lib/src/redis_client.dart:152]: This naming raised an eyebrow but seeing it in use is 💯

File: packages/redis_client/lib/src/redis_client.dart [hunk 4]
```diff
@@ -73,153 +141,169 @@ class RedisClient {
...
 
   /// Gets the value of a key.
   /// Returns null if the key does not exist.
   /// Equivalent to the `GET` command.
   /// https://redis.io/commands/get
-  Future<String?> get({required String key}) {
-    return _exec(() => RespCommandsTier2(_client!).get(key));
+  Future<String?> get({required String key}) async {
+    return await execute(['GET', key]) as String?;
   }
```
Comment (bryanoltman) [inline packages/redis_client/lib/src/redis_client.dart:183]: Fat arrows here

File: packages/redis_client/lib/src/redis_client.dart [hunk 5]
```diff
@@ -260,21 +376,17 @@ class RedisJson {
...
   /// https://redis.io/commands/json.get
   Future<Map<String, dynamic>?> get({required String key}) async {
-    final result = await _client.sendCommand([
-      'JSON.GET',
-      key,
-      r'$',
-    ]);
-    if (result is RespBulkString) {
-      final parts = LineSplitter.split(result.payload ?? '');
+    final result = await _client.execute(['JSON.GET', key, r'$']);
```
Comment (bryanoltman) [inline packages/redis_client/lib/src/redis_client.dart:388]: Thoughts on extracting these json method names to consts?

File: packages/redis_client/lib/src/redis_client.dart [hunk 6]
```diff
@@ -287,11 +399,23 @@ class RedisJson {
...
 }
 
 extension on RedisSocketOptions {
   /// The connection URI for the Redis server derived from the socket options.
   Uri get connectionUri => Uri.parse('redis://$host:$port');
 }
+
+final class _NoopRedisLogger implements RedisLogger {
+  const _NoopRedisLogger();
+  @override
```
Comment (bryanoltman) [inline packages/redis_client/lib/src/redis_client.dart]:
```suggestion
  const _NoopRedisLogger();

  @override
```

File: packages/redis_client/lib/src/redis_client.dart [hunk 7]
```diff
@@ -287,11 +399,23 @@ class RedisJson {
...
+    return _client.execute(['JSON.DEL', key, r'$']);
   }
 }
 
 extension on RedisSocketOptions {
   /// The connection URI for the Redis server derived from the socket options.
   Uri get connectionUri => Uri.parse('redis://$host:$port');
 }
+
+final class _NoopRedisLogger implements RedisLogger {
```
Comment (bryanoltman) [inline packages/redis_client/lib/src/redis_client.dart:412]: Just to make sure I'm not missing anything: this is the only `RedisLogger` implementation at the moment?

File: packages/redis_client/example/main.dart [hunk 8]
```diff
@@ -9,24 +7,9 @@ Future<void> main() async {
...
-  final value = await client.get(key: key); // WORLD
-  assert(value == 'WORLD', 'Value should be "WORLD".');
-
-  // Delete the key.
-  await client.delete(key: key);
-
-  final finalValue = await client.get(key: key); // null
-  assert(finalValue == null, 'Key should not exist.');
+  // Execute a command.
+  await client.execute(['PING']); // PONG
```
Comment (bryanoltman) [inline packages/redis_client/example/main.dart:12]: Why get rid of the get/set/delete examples?

**Review by bryanoltman — COMMENTED** [review]
A few small comments, but mostly looks good. I definitely think this could be its own package. Maybe `dart_redis` or something so people don't think it's related to code push in some way?

File: packages/redis_client/lib/src/redis_client.dart (hunk 4 above)
Comment (felangel) [reply packages/redis_client/lib/src/redis_client.dart:183]: it wraps weirdly 😛

File: packages/redis_client/lib/src/redis_client.dart (hunk 7 above)
Comment (felangel) [reply packages/redis_client/lib/src/redis_client.dart:412]: It's the only one that ships with the package but we'll create our own.

File: packages/redis_client/example/main.dart (hunk 8 above)
Comment (felangel) [reply packages/redis_client/example/main.dart:12]: Just thought the simplest getting started would be best but happy to change it back 🤷

File: packages/redis_client/lib/src/redis_client.dart (hunk 5 above)
Comment (felangel) [reply packages/redis_client/lib/src/redis_client.dart:388]: They should only be used in this one spot so I didn't see as much value but happy to do so if you prefer.

File: packages/redis_client/lib/src/redis_client.dart (hunk 3 above)
Comment (felangel) [reply packages/redis_client/lib/src/redis_client.dart:152]: haha thanks glad you thought so!

**Review by bryanoltman — APPROVED** [review]

**Pull request merged**